privacyidea/lib/eventhandler/scripthandler.py

```diff
@@ -164,14 +180,26 @@ def do(self, action, options=None):
             proc_args.append("--logged_in_role")
             proc_args.append(logged_in_user.get("role", "none"))
 
-        try:
-            p = subprocess.Popen(proc_args, cwd=self.script_directory)
-            log.info("Started script {script!r}:"
-                     " {process!r}".format(script=script_name, process=p))
-        except Exception as e:
-            log.warning("Failed to execute script {0!r}: {1!r}".format(
-                script_name, e))
-            log.warning(traceback.format_exc())
+        if handler_options.get("background", SCRIPT_BACKGROUND) == SCRIPT_BACKGROUND:
+            try:
+                p = subprocess.Popen(proc_args, cwd=self.script_directory)
+                log.info("Started script {script!r} in background:"
+                         " {process!r}".format(script=script_name, process=p))
+            except Exception as e:
+                log.warning("Failed to execute script {0!r}: {1!r}".format(
+                    script_name, e))
+                log.warning(traceback.format_exc())
+
+        elif handler_options.get("background") == SCRIPT_WAIT:
+            try:
+                log.info("Starting script {script!r} in foreground.".format(script=script_name))
+                r = subprocess.check_call(proc_args, cwd=self.script_directory)
+            except Exception as e:
+                log.warning("Failed to execute script {0!r}: {1!r}".format(
+                    script_name, e))
+                log.warning(traceback.format_exc())
+                if handler_options.get("raise_error"):
+                    raise e
```

Comment:
I think this is mergable, but just wanted to point out potential for code deduplication here: We could get rid of the outer if-branch and instead use
```python
            try:
                p = subprocess.Popen(proc_args, cwd=self.script_directory)
                log.info("Started script {script!r}:"
                         " {process!r}".format(script=script_name, process=p))
                if handler_options.get("background") == SCRIPT_WAIT:
                    log.info("waiting for script to finish ...")
                    ret = p.wait()
                    # handle error case if ret != 0
                    ...
            except Exception as e:
                log.warning("Failed to execute script {0!r}: {1!r}".format(
                    script_name, e))
                log.warning(traceback.format_exc())
```

What do you think?

Reply:
I was not aware of p.wait(). The question is, if it also raises an error accordingly?
Then we would - if ``raise_error`` was configured - have to reraise the error.

Reply:
You're right, we would need to re-raise an error. We could, however, just raise a privacyIDEA ``ServerError`` then.

Reply:
The problem is, that different situations would be handled differently. ``p.wait()`` does not raise an exception in case of an exit code <> 0.

The next problem is, that for some reason the tests do not work anymore - since the ``fail.sh`` script somehow does not work :-/